iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java

```diff
@@ -2005,39 +1986,47 @@ private void deleteDataInFiles(
       }
 
       ModificationFile modFile = tsFileResource.getModFile();
+      long originSize = modFile.getSize();
       if (tsFileResource.isClosed()) {
-        // delete data in sealed file
-        if (tsFileResource.isCompacting()) {
-          // we have to set modification offset to MAX_VALUE, as the offset of source chunk may
-          // change after compaction
-          deletion.setFileOffset(Long.MAX_VALUE);
-          // write deletion into compaction modification file
-          tsFileResource.getCompactionModFile().write(deletion);
-          // write deletion into modification file to enable read during compaction
-          modFile.write(deletion);
-          // remember to close mod file
-          tsFileResource.getCompactionModFile().close();
-          modFile.close();
-        } else {
-          deletion.setFileOffset(tsFileResource.getTsFileSize());
-          // write deletion into modification file
-          boolean modFileExists = modFile.exists();
-          long originSize = modFile.getSize();
-          modFile.write(deletion);
+        try {
+          tsFileResource.writeLock();
```

Comment:
```suggestion
          tsFileResource.writeLock();
          originSize = modFile.getSize();
```
We need to get file length inside lock, otherwise, there will be concurrent problems. ThreadB may truncate ThreadA's complete mod entry.